### PR TITLE
feat: visitor journey list

### PR DESCRIPTION
**Pull request by csiyang**
# Description

### <samp>🤖 Generated by Copilot at 9ac5d82</samp>

This pull request adds a new feature to the journeys-admin app: a visitors report page that shows the details of the visitors who interacted with a journey. It creates a new page component, a visitors list component, and a visitor card component with subcomponents for the header and details. It also adds some fake data and a translation key for the page title.

[- Link to Basecamp Todo](https://3.basecamp.com/3105655/buckets/32387047/todos/6067125896)

# How should this PR be QA Tested?

Please describe the QA tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Should show `No Visitors` if there are no visitors as well as hide the `Visitor Count` and `Load More` button
- [ ] Should show the current amount of displayed visitors against the total number of visitors (`Visitor Count`)
- [ ] Should display `VisitorCount` in the App Header
- [ ] Should load 100 visitors by default
- [ ] `Load More` button should load 100 more visitors
- [ ] `Load More` should be disabled if there are no more visitors to load
- [ ]  Should show a loading skeleton while the data is loading. 
- Visitor Card Header
> - [ ] Shows status, or person icon if by default if none is selected
> - [ ] Should show name as the last 12 characters of the user id, or the name the user entered in the `sign up` block
> - [ ] Should show location
> - [ ] Should show referrer
> - [ ] Should show journey started at time
> - [ ] Should show total duration spent on a journey
- Video Card Details
> - [ ] Should show name
> - [ ] Should show location
> - [ ] should not show any details if there are no events from the visitor
> - [ ] Should show all `ChatEvent`s from visitor for journey with red text
> - [ ] Should show all `TextResponseEvent`s from visitor for journey
> - [ ] Should show all `RadioSubmissionEvent`s from visitor for journey

# Walkthrough

### <samp>🤖 Generated by Copilot at 9ac5d82</samp>

*  Add a new page component for the visitors report of a journey ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-e9e60ec48f7928f205dea4cd73b81cd0498ead851b3bf03af91d9a24b3055ba8R1-R102))
* Add a new component for the visitors list of a journey ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-9ab5a5278961ae7639ad8a3e4d2e000affed870f0cb97975cad4efa824e70219R1-R18))
  * Render a list of visitor cards using some fake data for now ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-a7211f9b51edee7965b2e5ab102790e3748a2a5845dbba001fc837a0b8bddc2dR1-R87))
* Add a new component for the visitor card of a journey ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-a644effba72df05db723e52ab10085af1bc089be669052c76472904b5564736eR1-R50))
* Add a new component for the visitor card header of a journey ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-783805de11f0ecbfa7568ce88d392b2faf9358f23595dae6e9b640352de2aa16R1-R96))
* Add a new component for the visitor card details of a journey ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-d19a099f58b80b8252eaa566a2264b411c1b0d95d61cb46aba8514749dd51310R1-R64))
* Add index files to export the visitors list, visitor card, visitor card header, and visitor card details components ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-f5e09a462cf8416090d539dd498a351ea535f66d04e7a1a3bc1b581ce10ceee2R1), [link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-bad7e9a57d452d10cbd40457e9901a3e9a1a006487e803e47570ba68d7b10386R1), [link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-7223ee1a9a8bd10722a3563ff092913690ac3a830192848391d1b156d2657a71R1), [link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-72244d09f9ff946b8a416383a6b8f3fcd63c990a6dc325222c923fc1fafb1555R1))
* Add a new translation key and value for the visitors page title in the English locale file ([link](https://github.com/JesusFilm/core/pull/1541/files?diff=unified&w=0#diff-acd82a7e9a7b8b41e5d2ba5351ff70551bf8cb53d9a2dd9d010acedae21d2a2dR15))

### CODE REVIEW

**Review by mikeallisonJS — APPROVED** [review]

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        createdAt: '2021-11-19T12:35:56.647Z',
+        label: 'button label',
+        value: 'button value'
+      }
+    ]
+  }
+
+  it('should link to visitor details', () => {
+    const { getByRole, getAllByText } = render(
+      <VisitorCard visitorNode={visitorNode} />
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx:44]: lint issue

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      visitor: {
+        __typename: 'Visitor',
+        name: null,
+        status: null,
+        referrer: null
+      } as unknown as Visitor,
+      events: []
+    }
+
+    const { getAllByText } = render(<VisitorCard visitorNode={emptyVisitor} />)
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx:64]: lint issue

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          />
+        ) : status != null && !loading ? (
+          <Typography sx={{ mr: 3, display: { xs: 'none', sm: 'flex' } }}>
+            {status}
+          </Typography>
+        ) : (
+          <PersonOutlineRoundedIcon
+            sx={{ mr: 3, display: { xs: 'none', sm: 'flex' } }}
+          />
+        )}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx:53]: After looking at this some more, I think its confusing that the status icons live in different components for desktop and mobile, we should try and keep them in the same component. Lets move this back into the `VisitorCardHeader`

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        <Stack direction="column" sx={{ width: '100%' }}>
+          <VisitorCardHeader
+            loading={loading}
+            icon={visitorNode?.visitor.status ?? null}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx:57]:
```suggestion
            icon={visitorNode?.visitor.status}
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            name={
+              visitorNode?.visitor?.name ??
+              `#${visitorNode?.visitorId.slice(-12) as unknown as string}`
+            }
+            location={visitorNode?.countryCode ?? null}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx:62]:
```suggestion
            location={visitorNode?.countryCode}
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            source={visitorNode?.visitor.referrer ?? null}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx:63]:
```suggestion
            source={visitorNode?.visitor.referrer}
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            createdAt={visitorNode?.createdAt}
+            duration={visitorNode?.duration ?? null}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.tsx:65]:
```suggestion
            duration={visitorNode?.duration}
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        __typename: 'TextResponseSubmissionEvent',
+        id: 'text.id',
+        createdAt: '2021-11-19T12:35:56.647Z',
+        label: 'text label',
+        value: 'text value'
+      }
+    ]
+  }
+}
+
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.stories.tsx:73]: Add a story for loading

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    expect(getAllByText('#012345678901')).toHaveLength(2)
+  })
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx:66]: add a test for skeleton

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    return {
+      t: (str: string) => str
+    }
+  }
+}))
+
+describe('VisitorCardDetails', () => {
+  it('should show name', () => {
+    const { getByText } = render(
+      <VisitorCardDetails
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx:17]: lint

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+      {
+        __typename: 'ButtonClickEvent',
+        id: 'ButtonClickEvent.id',
+        createdAt: 'isostring',
+        label: 'button label',
+        value: 'button value'
+      }
+    ]
+    const { getByText, queryByText } = render(
+      <VisitorCardDetails name="test name" events={events} />
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx:65]: lint

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    const props = {
+      icon: null,
+      name: '12345234',
+      location: null,
+      source: null,
+      createdAt: '2021-11-19T12:34:56.647Z',
+      duration: 45
+    }
+
+    const { getAllByTestId } = render(<VisitorCardHeader {...props} />)
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.spec.tsx:44]: lint

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+function DetailsRow({
+  label,
+  value,
+  chatEvent = false,
+  loading
+}: DetailsRowProps): ReactElement {
+  const textColor = chatEvent ? 'primary' : 'secondary'
+  const isBold = chatEvent ? 900 : 'normal'
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx:79]: Remove
```suggestion
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  const variant = chatEvent ? 'subtitle1' : 'body2'
+  return (
+    <>
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx:82]: Fragment not needed
```suggestion
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    )
+
+    expect(getByText('Chat Started')).toBeInTheDocument()
+    expect(getByText('text value')).toBeInTheDocument()
+    expect(getByText('radio value')).toBeInTheDocument()
+    expect(queryByText('button value')).not.toBeInTheDocument()
+  })
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.spec.tsx:72]: add test for loading

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    expect(getAllByTestId('PersonOutlineRoundedIcon')).toHaveLength(2)
+  })
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.spec.tsx:46]: add test for loading

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+function DetailsRow({
+  label,
+  value,
+  chatEvent = false,
+  loading
+}: DetailsRowProps): ReactElement {
+  const textColor = chatEvent ? 'primary' : 'secondary'
+  const isBold = chatEvent ? 900 : 'normal'
+  const variant = chatEvent ? 'subtitle1' : 'body2'
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardDetails/VisitorCardDetails.tsx:80]: Try subtitle2? Subtitle 1 is a bit too much
```suggestion
  const variant = chatEvent ? 'subtitle2' : 'body2'
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/JourneyVisitorsList.tsx [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  loading,
+  hasNextPage = false
+}: Props): ReactElement {
+  const { t } = useTranslation('apps-journeys-admin')
+  const hasVisitors = visitorEdges != null && visitorEdges.length > 0
+  return (
+    <Stack
+      spacing={6}
+      alignItems="center"
+      sx={{ width: '100%', height: '100%' }}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/JourneyVisitorsList.tsx:33]: Add this to remove the margins for mobile
```suggestion
      sx={{ width: '100%', height: '100%' , mx: { xs: -6, sm: 0 }}}
```

**Comment by csiyang** [comment]
Also check the related stories, they will need to be updated since props have changed

File: apps/journeys-admin/pages/journeys/[journeyId]/reports/visitors.tsx [hunk 19]
```diff
@@ -19,16 +20,113 @@ import i18nConfig from '../../../../next-i18next.config'
...
+  const [visitorEdges, setVisitorEdges] = useState<VisitorEdge[]>([])
+  const [hasNextPage, setHasNextPage] = useState(false)
+  const [endCursor, setEndCursor] = useState<string | null>()
+
+  const { fetchMore, loading } = useQuery<GetJourneyVisitors>(
+    GET_JOURNEY_VISITORS,
+    {
+      variables: {
+        filter: { journeyId },
+        first: 50
```
Comment (csiyang) [inline apps/journeys-admin/pages/journeys/[journeyId]/reports/visitors.tsx:103]: Steven has asked for this to be 100
```suggestion
        first: 100
```

File: apps/journeys-admin/pages/journeys/[journeyId]/reports/visitors.tsx [hunk 20]
```diff
@@ -19,16 +20,113 @@ import i18nConfig from '../../../../next-i18next.config'
...
+      }
+    }
+  )
+
+  async function handleFetchNext(): Promise<void> {
+    if (hasNextPage) {
+      const response = await fetchMore({
+        variables: {
+          filter: { journeyId },
+          first: 50,
```
Comment (csiyang) [inline apps/journeys-admin/pages/journeys/[journeyId]/reports/visitors.tsx:118]:
```suggestion
          first: 100,
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+      events: []
+    }
+
+    const { getAllByText } = render(
+      <VisitorCard visitorNode={emptyVisitor} loading={false} />
+    )
+    expect(getAllByText('#012345678901')).toHaveLength(2)
+  })
+
+  it('it should show skeleton when loading', () => {
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCard.spec.tsx:70]:
```suggestion
  it('should show skeleton when loading', () => {
```

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.tsx [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+          {source}
+        </Typography>
+        <Stack direction="row" sx={{ ml: 'auto' }}>
+          {createdAt != null && !loading ? (
+            <Typography
+              variant="subtitle1"
+              noWrap
+              sx={{ color: 'secondary.light' }}
+            >
+              {format(parseISO(createdAt), 'h:mmaaa, LLL. do')}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.tsx:90]: Update to use intl format

File: apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.tsx [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+        ) : status != null ? (
+          <Typography sx={{ mr: 3, mt: 6 }}>{status}</Typography>
+        ) : (
+          <PersonOutlineRoundedIcon sx={{ mr: 3, mt: 6 }} />
+        )}
+        <Stack direction="column">
+          <Stack direction="row">
+            {createdAt != null && !loading ? (
+              <Typography variant="subtitle1" sx={{ color: 'secondary.light' }}>
+                {format(parseISO(createdAt), 'h:mmaaa, LLL. do')}
```
Comment (csiyang) [inline apps/journeys-admin/src/components/JourneyVisitorsList/VisitorCard/VisitorCardHeader/VisitorCardHeader.tsx:130]: Update to use intl format

**Review by mikeallisonJS — APPROVED** [review]

**Pull request merged**